applicationset/services/scm_provider/azure_devops_test.go

```diff
@@ -0,0 +1,287 @@
+package scm_provider
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+
+	azureMock "github.com/argoproj/argo-cd/v2/applicationset/services/scm_provider/azure_devops/git/mocks"
+	"github.com/microsoft/azure-devops-go-api/azuredevops"
+	azureGit "github.com/microsoft/azure-devops-go-api/azuredevops/git"
+)
+
+type uuidHolder struct {
+	uuid uuid.UUID
+	err  error
+}
+
+func getUuidHolder() uuidHolder {
+	uuid, err := uuid.NewUUID()
+	return uuidHolder{uuid: uuid, err: err}
+}
+
+func s(input string) *string {
+	return &input
+}
+
+func TestAzureDevopsRepoHasPath(t *testing.T) {
+	organization := "myorg"
+	teamProject := "myorg_project"
+	repoName := "myorg_project_repo"
+	path := "dir/subdir/item.yaml"
+	branchName := "my/featurebranch"
+
+	ctx := context.TODO()
+	uuidHolder := getUuidHolder()
+
+	testCases := []struct {
+		name             string
+		pathFound        bool
+		azureDevopsError error
+		returnError      bool
+	}{
+		{
+			name:      "repo_has_path_found_returns_true",
+			pathFound: true,
+		},
+		{
+			name:             "repo_has_path_not_found_empty_search_result_returns_false",
+			pathFound:        false,
+			azureDevopsError: azuredevops.WrappedError{TypeKey: s("GitItemNotFoundException")},
+		},
+		{
+			name:             "repo_has_path_not_found_other_azure_devops_error_returns_error",
+			pathFound:        false,
+			azureDevopsError: azuredevops.WrappedError{TypeKey: s("OtherAzureDevopsException")},
+			returnError:      true,
+		},
+		{
+			name:             "repo_has_path_not_found_other_error_returns_error",
+			pathFound:        false,
+			azureDevopsError: fmt.Errorf("Undefined error from Azure Devops"),
+			returnError:      true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			gitClientMock := azureMock.Client{}
+
+			clientFactoryMock := &AzureClientFactoryMock{mock: &mock.Mock{}}
+			clientFactoryMock.mock.On("GetClient", mock.Anything).Return(&gitClientMock, nil)
+
+			repoIdAsString := fmt.Sprintf("%v", uuidHolder.uuid)
+			gitClientMock.On("GetItem", ctx, azureGit.GetItemArgs{Project: &teamProject, Path: &path, VersionDescriptor: &azureGit.GitVersionDescriptor{Version: &branchName}, RepositoryId: &repoIdAsString}).Return(nil, testCase.azureDevopsError)
+
+			provider := AzureDevopsProvider{organization: organization, teamProject: teamProject, clientFactory: clientFactoryMock}
+
+			repo := &Repository{Organization: organization, Repository: repoName, RepositoryId: uuidHolder.uuid, Branch: branchName}
+			hasPath, err := provider.RepoHasPath(ctx, repo, path)
+
+			if testCase.returnError {
+				assert.Error(t, err)
```

Comment:
curious! why we have this as a check? both the scenario error or noerror as passing test case?

Reply:
Probably we should segregate it with two names